# <span style="color: #2BF507; background-color: #024E4C; padding: 15px">Roteiro 1 - Introdução à Linguagem Python </span> 

## Conteúdo
 - Introdução à Linguagem Python - Sintaxe Básica e Pacotes NumPy e Matplotlib

## Programação


<h2 style="background-color: #09C8C1; color: #07199C; padding: 15px">
    <b>Introdução à Linguagem Python</b>
</h2>

### 1 - Instalar interpretador **Python** e servidor **Jupyter Notebook**

Python usa um interpretador que pode ser executado em uma janela de terminal (**cli** - *commands line interface*), permitindo que se experimente *snippets* (pequenos trechos de programa em Python) sem precisar salvar e executar um programa inteiro.  

#### Instalação no **Windows**  
O Windows, geralmente vem sem o interpretador Python, então vamos baixá-lo e instalá-lo.

Primeiro, verifique se o Python está instalado no seu sistema operacional S.O.). Abra uma janela de comando (janela de **terminal**) e digite `python` com letras minúsculas. Se você receber um *prompt* do Python (<span style="color: #2BF507;"><b> >>> </b></span>), então o Python está instalado em seu S.O.  
Caso contrário, será mostrada uma mensagem de erro informando que `python` não é um comando reconhecido pelo sistema. Neste caso, baixe um instalador Python direto do site oficial`http://python.org/downloads`, escolhendo a versão mais atualizada. Execute o arquivo baixado, não deixando de marcar a opção `Add Python to PATH`.

<img src="rot1_fig1.png" alt="Tela do instalador Python no Windows" width="400" style="display: block; margin: 0 auto;">  

<p><p>Abra uma janela de comando e digite python em letras minúsculas. Se você receber um prompt do Python (<span style="color: #2BF507;"><b> >>> </b></span>), o Windows encontrou a versão do Python que você acabou de instalar:</p>

``` 
c:\> python
Python 3.12.3 (v3.12.3:238974d21fe, Oct 16 2024, 08:11:21) [MSC v.1900 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>>
```
Ao carregar o interpretador, um prompt com três sinais de maior (<span style="color: #2BF507;"><b> >>> </b></span>) indicará que você pode digitar comandos em Python para serem interpretados.

### 2 - Manipulando Dados em Python: Números, Textos e Variáveis

Manipular dados é fundamental para qualquer projeto de programação, desde *scripts* simples até análise de dados complexa. A ling. Python facilita esse processo com uma sintaxe clara e tipos de dados dinâmicos, onde a manipulação se baseia em números, textos (strings) e variáveis.

#### 2.1 Variáveis em Python
As variáveis são armazenadas localmente na memória RAM do computador.  

* **Tipagem Dinâmica**: Não é preciso declarar o tipo da variável antes de usá-la, assim como ocorre em outras linguagens; o interpretador Python a define automaticamente.
* **Declaração**: ` idade = 25, nome = "Ana"`.
* **Reatribuição**: O valor de uma variável pode mudar seu valor e seu tipo durante a execução do código.
* **Função `type()`**: Retorna o tipo de dado de uma variável.

In [1]:
idade = 35
nome = "João"
print("Nome: ", nome,"\nIdade: ", idade)
print(type(nome), type(idade))

Nome:  João 
Idade:  35
<class 'str'> <class 'int'>


#### 2.2 Manipulação de Números
Python gerencia dois tipos principais de números:
* **Inteiros (int)**: Números inteiros não possuem casa(s) decimal(is); `10, -5`
* **Ponto Flutuante (float)**: Números com parte fracionária; `-10.5, 3.1415926535`
* **Operações Aritméticas**: Soma (+), Subtração (-), Multiplicação (*), Divisão (/), Divisão Inteira (//), Resto da divisão (%) e Potenciação (**)
* **Formatação**: Podemos usar o formato `.2f` para limitar em duas casas decimais a exibição do valor de uma variável do tipo `float`

##### 2.2.1 Formatação F-string
As `F-strings` oferecem a sintaxe mais legível para formatações: `f"{objeto:especificador_de_formato}"`.   
Permitem definir casas decimais (`.2f`), alinhar texto, adicionar símbolos de moeda (R$) ou porcentagem (%) e formatar datas.

In [ ]:
print(f"{-10.5:.5f}\nPi ={3.1415926535:10.3f}")  # formatando valores reais (ponto flutuante)
print(f"Valor da compra : R$ {1230.55:.2f}")     # formatando valor monetário

from datetime import datetime   # importação do objeto 'datetime'
agora = datetime.now()          # retorna a data e o horário correntes
# Formata como DD/MM/YYYY
print(f"Data de Abertura: {agora:%d/%m/%Y}")     # formatando valor de data

-10.50000
Pi =     3.142
Valor da compra : R$ 1230.55
Data de Abertura: 28/04/2026


#### 2.3 Manipulação de Textos (*Strings*)
*Strings* são sequências de caracteres delimitadas por aspas simples (' ') ou duplas (" "). 
* **Concatenação**: Une *strings* usando o operador '+'; `nome + " Silva"`
* **Acesso** e **Fatiamento (*Slicing*)**: Acessar partes do texto: `texto[0]` pega a primeira letra, `texto[2:5]` pega da terceira à quinta letra
* **Métodos Importantes**  
    `len()`: Retorna a quantidade de caracteres da string.  
    `strip()`: Remove espaços em branco no início e final da string.  
    `upper()` / `lower()`: Converte para maiúsculas ou minúsculas os caracteres alfabéticos da string.  
    `split()`: Divide a string em uma lista de strings (separador padrão é o espaço em branco).  
    `find()`: Localiza uma substring dentro da string.  

In [7]:
nome = "João"                               # variável do tipo string '<str>'
sobrenome = "Silva"                         # variável do tipo string '<str>'
print(nome+sobrenome)                       # concatenação de strings
print(nome+" "+sobrenome)                   # concatenação de strings 
print("Sigla:", nome[0]+sobrenome[0])       # concatenação de caracteres
print(nome.upper(), sobrenome.upper())      # conteúdo da string em minúsculas e maiúsculas
print("Posição da substring 'Si':", (nome+" "+sobrenome).find('Si'))

JoãoSilva
João Silva
Sigla: JS
JOÃO SILVA
Posição da substring 'Si': 5


In [4]:
nome = "Python"
versao = 3.12
# Usando f-string para formatar texto e números:
mensagem = f"Bem-vindo ao {nome} na versão {versao}"
print(mensagem.upper()) # Saída: BEM-VINDO AO PYTHON NA VERSÃO 3.1

BEM-VINDO AO PYTHON NA VERSÃO 3.12


#### 2.4 Conversão de Tipos (*Type Casting*)
É possível converter dados para formatos diferentes para garantir que a operação funcione corretamente: 
* `int()  `: Converte para inteiro; `int("10") --> 10`
* `float()`: Converte para número decimal; `float("10.5") --> 10.5`
* `str()  `: Converte números para texto;  `str(-25.1) --> "-25.1"`  

In [13]:
valor = "1500.00"
print("Valor      : R$", valor)
# desconto de 10%
desc = float(valor) * 0.9
print(f"Valor final: R$ {desc:.2f}")

Valor      : R$ 1500.00
Valor final: R$ 1350.00


### 3 - Estruturas de Dados Nativas

Tratam-se de estruturas de dados nativas para armazenagem de __dados ordenados__ em Python.  
As **Listas** são mutáveis (alterações, acréscimos ou remoções de itens são permitidos), ideais para armazenamento de coleções dinâmicas de dados.  
As **Tuplas** são imutáveis (não podem ser alteradas, acrescidas ou removidas após a criação), sendo mais rápidas, seguras para armazenamento de dados constantes e ideais para registros.

#### 3.1 Listas e Tuplas
* **Listas** (list):

    Exemplo: `minha_lista = [1, 2, "python", 3.14, True]`  
    Operações: `.append()` para anexação de itens, `.remove()` para retirada de itens, ou alterações via uso de índice do item: `lista[0] = 10`  
    Uso: Coleções de dados heterogêneos que mudam de tamanho ou valor frequentemente

* **Tuplas** (tuple):

    Exemplo: `minha_tupla = (1, 2, "python", 3.14, True)`  
    Operações: por serem estruturas imutáveis, uma vez criadas, não podem ser alteradas  
    Uso: Dados estruturais, constantes, retornos de funções (mais rápido e seguro)  

* **Comportamentos** comuns a ambas:  
    Ordenadas: Mantêm a ordem de inserção dos dados 
    Indexação: Acessíveis por índices específicos [0], [-1], ou por fatiamento [1:3], [:4]
    Heterogêneas: Podem conter tipos de dados diferentes (inteiros, strings etc.)

    Use **listas** quando precisar alterar o conteúdo ao longo da execução do código.  
    Use **tuplas** para garantir a integridade dos dados e otimizar o desempenho quando a estrutura não precisar de alterações. 

In [ ]:
# Exemplo 1: uso de lista
lista_teste = ['Engenheiro', 'Médico', 'Arquiteto','Professor']
print(type(lista_teste))
print(lista_teste)
print(lista_teste[0])
print(lista_teste[1])
print(lista_teste[2:4])

<class 'list'>
['Engenheiro', 'Médico', 'Arquiteto', 'Professor']
Engenheiro
Médico
['Arquiteto', 'Professor']


In [ ]:
# Exemplo 2: modificação do último elemento da lista
lista_teste = ['Engenheiro', 'Médico', 'Arquiteto', 'Engenheiro']  # elementos repetidos na lista
print(type(lista_teste), lista_teste)
print(lista_teste[0])
print(lista_teste[1])
print(lista_teste[2:4])
lista_teste[3] = 'Farmacêutico'    # alteração do último elemento da lista
print(lista_teste)

<class 'list'> ['Engenheiro', 'Médico', 'Arquiteto', 'Engenheiro']
Engenheiro
Médico
['Arquiteto', 'Engenheiro']
['Engenheiro', 'Médico', 'Arquiteto', 'Farmacêutico']


In [19]:
# Exemplo 3: anexação de um elemento à lista
outra_lista = ['Engenheiro', 10, 10.5, 'Médico', -1]
print(outra_lista)
print(outra_lista[0])
print(outra_lista[3])
print(outra_lista[4])
print(outra_lista[0:4])
outra_lista.append(1001)     # anexando um novo elemento à lista
print(outra_lista)

['Engenheiro', 10, 10.5, 'Médico', -1]
Engenheiro
Médico
-1
['Engenheiro', 10, 10.5, 'Médico']
['Engenheiro', 10, 10.5, 'Médico', -1, 1001]


In [ ]:
# Exemplo 4: uso de tuplas
tupla1 = (23, "ok")            # criação de uma tupla
tupla2 = tuple(outra_lista)    # conversão de lista em tupla
print(tupla1, tupla2)
print(tupla1[0])
# tupla1[0] = 25                 # retire o comentário e execute a célula
# tupla1.append(1001)            # retire o comentário e execute a célula

(23, 'ok') ('Engenheiro', 10, 10.5, 'Médico', -1, 1001)
23


#### 3.2 Pilhas e Filas
Python não possui estruturas nativas para **Pilhas** ou **Filas**, embora possamos implementá-las usando **listas**.  

**Pilha**  
Pilha é uma lista de elementos onde novos elementos são adicionados somente no final (topo) da lista, ou removidos somente do final (topo) da lista. 
Numa pilha de livros, sempre que adicionamos livro(s) a pilha cresce, e sempre que removemos livro(s) a pilha diminui. Essa estrutura de dados é baseado no princípio **"último a entrar, é o primeiro a sair"** (LIFO - *Last-In First-Out*).

In [ ]:
pilha = [1, 2, 3, 4]  # pilha construída a partir de uma lista
print(pilha)
pilha.append(5)       # acrescenta item no topo da pilha
print(pilha)
pilha.pop()           # retira item do topo da pilha
print(pilha)

[1, 2, 3, 4]
[1, 2, 3, 4, 5]
[1, 2, 3, 4]


**Filas**  
Fila é uma lista de elementos na qual novos elementos são adicionados ao final da lista, e os elementos são removidos do início da lista. É comparável a ficar numa fila no mundo real. A fila fica mais curta à medida que os elementos que estão na frente são atendidos. Fica mais longa quando uma nova pessoa entra no final da fila. Essa estrutura de dados é baseado no princípio **"primeiro a entrar, é o primeiro a sair"** (FIFO - *First-In First-Out*).

In [ ]:
fila = [1, 2, 3, 4]
print(fila)
fila.append(5)        # acrescenta item no final da fila
print(fila)
fila.pop(0)           # retira item do início da fila
print(fila)

[1, 2, 3, 4]
[1, 2, 3, 4, 5]
[2, 3, 4, 5]


### 4 - Controlando o Fluxo de Execução: if / elif / else 

O controle de fluxo com `if, elif` e `else` em Python permite que o programa tome decisões, executando blocos de código específicos com base em condições booleanas (`True/False`). O `if` avalia uma condição, `elif`  (abreviação de `else if`) verifica condições adicionais e `else` define o comportamento se nenhuma condição anterior for verdadeira.

#### Sintaxe
* **Identação**: Python usa identação (espaços no início da linha) **compulsoriamente** para definir o bloco de código de cada condição.
* **Dois Pontos (:)**: As instruções `if, elif` e `else` devem terminar com dois pontos.
* **Operadores Relacionais**: Utilizados para comparar valores, `==` (igualdade), `!=` (diferente), `>`, `<`, `>=`, `<=`.
* **Operadores Lógicos**: Combinam expressões relacionais, `and` (e), `or` (ou), `not` (não).
* **If Curto** (Ternário): É uma estrutura condicional simples escrita numa única linha:  
`resultado = "Aprovado" if nota >= 7 else "Reprovado"`

In [23]:
# Exemplo de uso da estrutura condicional de fluxo de execução
nota = 7.5
if nota >= 9:
    print("Nota A")
elif nota >= 7:
    print("Nota B")
else:
    print("Nota C")

Nota B


In [ ]:
idade = 18
if idade < 13:
    print("Criança")
elif 13 <= idade < 18:
    print("Adolescente")
else:
    print("Adulto")

Adulto


### 5 - Entrada de Dados pelo Usuário (via Teclado)

A entrada de dados via teclado em Python é realizada principalmente pela função nativa `input()`, que pausa a execução do programa e aguarda o usuário digitar um valor, **retornando sempre uma *string*** (texto). É recomendável usar uma mensagem de *prompt* dentro da função para guiar o que o usuário deve inserir (digitar).

#### Características
* **Uso Básico**: `nome = input("Digite seu nome: ")`
* **Captura como String**: Todo dado é lido como texto. Se precisar de números, é preciso converter (*type casting*).
* **Conversão de Dados**

    **Inteiros**: `idade = int(input("Digite sua idade: "))`  
    **Ponto Flutuante**: `altura = float(input("Digite sua altura: "))`  
    **Comportamento**: O programa aguarda até que a tecla [Enter] seja pressionada.

In [29]:
nome = input("Qual é o seu nome? ")
prenome = nome.split()[0]
idade = int(input("Qual é a sua idade? "))
altura = float(input("Digite sua altura (m): "))
print(f"Olá {prenome}, você tem {idade} anos e estatura de {altura:.2f} m.")

Olá João, você tem 16 anos e estatura de 1.66 m.


### 6 - Controlando o Fluxo de Execução: while / for

O controle de fluxo em Python utiliza `for` para iterar sobre sequências (listas, tuplas, strings, range) um número definido de vezes, e `while` para repetir códigos enquanto uma condição for verdadeira (ideal para números de iterações desconhecidos). Ambos laços podem ser interrompidos por `break` (abandona iterações) e `continue` (força a próxima iteração).

#### Laço `for` (Quantidade de Repetições Definida)
O `for` em Python é usado para percorrer os itens de uma sequência na ordem em que aparecem.

In [ ]:
# Iterando sobre uma lista
salada = ["maçã", "banana", "uva"]
for fruta in salada:
    print(fruta)

maçã
banana
uva


In [ ]:
# Usando a função nativa range() para realizar 4 repetições (0 a 3)
for i in range(4):
    print(f"Passo {i}")

Passo 0
Passo 1
Passo 2
Passo 3


#### Laço `while` (Quantidade de Repetições Indefinida)
O `while` repete um bloco de código enquanto a condição especificada for `True`. É crucial garantir que a condição se torne `False` eventualmente para evitar laços infinitos.

In [30]:
# Contador de passos
contador = 0
while contador < 4:
    print(f"Passo {contador}")
    contador += 1      # Essencial para evitar laço infinito

Passo 0
Passo 1
Passo 2
Passo 3


#### Interrupção e Continuação de Laços
* `break`: Finaliza o laço imediatamente.
* `continue`: Ignora o restante do código na iteração atual e pula para a próxima iteração.

In [32]:
# break
for n in range(10):
    if n >= 4:
        break         # Para o loop quando n for 4
    print(f"Passo {n}")
print("Laço encerrado!")

Passo 0
Passo 1
Passo 2
Passo 3
Laço encerrado!


In [33]:
# continue
for n in range(5):
    if n == 2:
        continue      # Pula a impressão do número 2
    print(f"Passo {n}")

Passo 0
Passo 1
Passo 3
Passo 4


# Exercícios


## 1)Use dos conhecimentos adiquiridos para:
a) Crie variáveis e atribua valores para nome, idade(anos), altura(cm) e cidade de moradia de três pessoas e imprima as mesmas.

b) Escreva agora as alturas do item a) em metros usando `float` e a mostre com duas casas após a virgula e imprima.

c) Crie uma fila com as alturas das pessoas do item b), adicione mais dois valores quaisquer e imprima.

d) Pegue o primeiro elemento da lista e avalie se ele é maior que 1.60, se for imprima o valor, se não imprima a mensagem "altura abaixo do esperado!".

e) Crie um programa para atribuir as quatro variaveis do item a) usando `input` para três pessoas e imprima as mesmas.

Resposta das questoes B e D

In [ ]:
B
altura = float(input("1.69", "1.90", "1.54"))
print(f"{altura:.2f} m.")
D
altura = [1.69, 1.90, 1.54]
if  altura[0]>= 1.60:
    print(altura[0])
else:
    print("altura abaixo do esperado!")